compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/CollectionStubMethodLowering.kt

```diff
@@ -0,0 +1,274 @@
+        // our stubs based on these functions. For every collection class we select specific function(s), to make sure all
+        // type parameters are covered by our selection.
+        val targetTypeList = when (collectionClass.name.asString()) {
+            "Iterator", "ListIterator" -> {
+                listOf("next")
+                    .map { functionName -> getFunctionForSubstitution(functionName, collectionClass, targetClass) }
+                    .map { function -> function.returnType }
+            }
+            "Entry" -> {
+                listOf("<get-key>", "<get-value>")
+                    .map { functionName -> getFunctionForSubstitution(functionName, collectionClass, targetClass) }
+                    .map { function -> function.returnType }
+            }
+            // Iterable is special, the only way to get type parameter is through iterator() function,
+            // which returns Iterator<T>, therefore we need to look further into type arguments to get the substitution type.
+            "Iterable" -> {
+                listOf("iterator")
+                    .map { functionName -> getFunctionForSubstitution(functionName, collectionClass, targetClass) }
+                    .flatMap { function ->
+                        (function.returnType as IrSimpleType).arguments
+                            .map { arg -> (arg as IrTypeProjection).type }
+                    }
+            }
+            "Collection", "List", "Set" -> {
+                listOf("contains")
+                    .map { functionName -> getFunctionForSubstitution(functionName, collectionClass, targetClass) }
+                    .map { function -> function.valueParameters[0].type }
+            }
+            "Map" -> {
+                listOf("containsKey", "containsValue")
+                    .map { functionName -> getFunctionForSubstitution(functionName, collectionClass, targetClass) }
+                    .map { function -> function.valueParameters[0].type }
+            }
+            else -> throw IllegalStateException()
+        }
+        if (targetTypeList.isNullOrEmpty() || targetTypeList.size != mutableClass.typeParameters.size) {
+            throw IllegalStateException()
+        }
+        return mutableClass.typeParameters.map { it.symbol }.zip(targetTypeList).toMap()
+    }
+
+    // we can write function with same amount of value arguments and same name as those defined in collection interface,
+    // therefore we need to check for overridden symbols to make sure we are looking at the correct one.
+    private fun getFunctionForSubstitution(functionName: String, collectionClass: IrClass, targetClass: IrClass) =
+        targetClass.functions
+            .filter { function -> function.name.asString() == functionName }
+            .filter { function -> overridesFunctionInImmutable(collectionClass, function) }
+            .first()
```

Comment:
We usually prefer to use less calls and lambdas if it doesn't make the code heavier or harder to understand. Also, instead of filtering something into a collection and then taking the first element, you can simply take the first element that satisfies the predicate:
```
        targetClass.functions.first { function ->
            function.name.asString() == functionName && overridesFunctionInImmutable(collectionClass, function)
        }
```

Reply:
Also, consider using `firstOrNull` instead of `first`, and adding `?: error("...")` with an error description mentioning the class and what happened